Update mio to 0.8

flickpp opened the pull request:
close #53 

changes:
  - update mio to 0.8 in Cargo.toml
  - mio.rs now uses newer 0.8 mio API

My implementation removes the dependency `fd_queue::mio::UnixStream` on `fd_queue::net::UnixStream`. Much like how `fd_queue::tokio::UnixStream` is implemented.

flickpp committed:
Update mio to 0.8

changes:
  - update mio to 0.8 in Cargo.toml
  - mio.rs now uses newer 0.8 mio API

sbosnick commented on Cargo.toml:
```diff
@@ -17,7 +17,7 @@ tokio-fd = ["tracing", "tokio", "pin-project", "futures-core", "futures-util"]
 [dependencies]
 tracing = {version = "0.1.15", optional = true}
-mio = {version = "0.6.22", optional = true}
+mio = {version = "0.8.4", optional = true, features = ["os-poll", "os-ext", "net"]}
```
I don't think feature "os-ext" is needed. If we don't need it then I would prefer to not include it here because features are additive so this would impose "os-ext" on users of the crate who may not need it.

Similarly please think about whether "os-poll" is needed. I am not as familiar with `mio` as I am with `tokio` so I will accept your judgement here, but it looks from the `mio` documentation that "os-poll" is needed to do anything useful with `mio` but it is not necessary to compile against `mio` as a dependency. In that case it may be better to omit "os-poll" here and let the user of the crate add it.

flickpp replied:
Yeah I think you're right.

Moved to dev-dependencies as both are needed for the test suite,

sbosnick commented on src/mio.rs:
```diff
@@ -73,67 +90,98 @@ impl UnixStream {
+/// The [`RawFd`][RawFd] that are received along with the bytes will be available
+/// through the method of the `DequeueFd` implementation. The number of
+/// [`RawFd`][RawFd] that can be received in a single call to one of the `Read`
+/// methods is bounded by `FD_QUEUE_SIZE`. It is an error if the other side of this
```
The reference to `FD_QUEUE_SIZE` here might be misleading. Could you please either add a const `FD_QUEUE_SIZE` to `fd_queue::mio::UnixStream` or change this reference to refer to the `net` implementation of `UnixStream`.

flickpp replied:
Thanks! Good catch.

sbosnick commented on src/mio.rs:
```diff
@@ -73,67 +90,98 @@ impl UnixStream {
+
+    /// Execute an I/O operation ensuring that the socket receives more events
+    /// if it hits a WouldBlock error.
+    /// See https://docs.rs/mio/latest/mio/net/struct.UnixStream.html#method.try_io
```
This line gives an error for `cargo doc --feature mio-fd` (which doesn't yet show up for CI).

flickpp replied:
Thanks! Good Catch.

sbosnick reviewed (changes requested):
Thank you for this pull request. It solves the problem in #53 well. I have listed a few small issues that I think should be addressed before merging this. Please also adjust the `mio-fd` feature in `Cargo.toml` to remove the reference to `net-fd` since it is no longer needed (this may require adding a reference to `tracing` there).

Can you also consider adding tests like the `unix_stream_passes_fd` and `unix_stream_connects_to_listener` tests from the Tokio implementation. When the `mio` implementation relied on the `net` implementation we could rely on the test from there to make sure that the basic functionality worked but with the separation implemented in this PR we can't do that any more. If you don't have time to add these test then it is fine to just add an issue referring to this so that I or someone else can get to it later.

flickpp commented:
Something else which I've found using mio (and this crate!) is that a UnixStream connected with bind doesn't delete the socket file on Drop.

I'd like to add _deleting_ the socket file when the UnixListener is dropped - but I do appreciate this is functionality different to mio.

flickpp committed:
Moved mio-ext, mio-poll to dev. New test.

Changes:
   - mio-ext and mio-poll are now only required by the test suite
   - new (currently failing) test passing an FD

flickpp commented on src/mio.rs:
```diff
@@ -366,6 +367,87 @@ mod tests {
     }
 
+    #[test]
+    fn stream_bind_and_connect_and_pass_open_file() {
```
I don't why this is failing. I'll have a look over the next week.

I can tell you, it works for my use case of passing tcp listeners between processes.
